File: kluctl/deployment/hooks_util.py
import dataclasses
import logging

from typing import List, Set, Optional

from kluctl.utils.dict_utils import get_dict_value
from kluctl.utils.k8s_object_utils import get_object_ref, get_long_object_name
from kluctl.utils.utils import parse_bool

logger = logging.getLogger(__name__)

@dataclasses.dataclass
class Hook:
    object: dict
    hooks: Set[str]
    weight: int
    delete_policies: Set[str]
    wait: bool

class HooksUtil:
    def __init__(self, apply_util):
        self.apply_util = apply_util

    def run_hooks(self, d, hooks):
        def do_log(level, str):
            self.apply_util.do_log(d, level, str)

        l = self.get_sorted_hooks_list(d.objects)
        l = [x for x in l if any(h in hooks for h in x.hooks)]

        delete_before_objects = []
        apply_objects = []

        for h in l:
            if self.apply_util.abort_signal:
                return
            if "before-hook-creation" in h.delete_policies:
                delete_before_objects.append(h)
            apply_objects.append(h)

        if len(delete_before_objects) != 0:
            do_log(logging.INFO, "Deleting %d hooks before hook execution" % len(delete_before_objects))
        for h in delete_before_objects:
            do_log(logging.DEBUG, "Deleting hook %s due to hook-delete-policy %s" % (get_long_object_name(h.object), ",".join(h.delete_policies)))
            self.apply_util.delete_object(get_object_ref(h.object))

        if len(apply_objects) != 0:
            do_log(logging.INFO, "Applying %d hooks" % len(delete_before_objects))
        for h in apply_objects:
            if self.apply_util.dry_run and "before-hook-creation" in h.delete_policies:
                self.apply_util.handle_result(h.object, [])
                continue
            do_log(logging.DEBUG, "Applying hook %s" % get_long_object_name(h.object))
            self.apply_util.apply_object(h.object)

        wait_results = {}
        for h in l:
            if self.apply_util.abort_signal:
                return
            ref = get_object_ref(h.object)
            if self.apply_util.had_error(ref):
                continue
            if not h.wait:
                continue
            wait_results[ref] = self.apply_util.wait_object(ref)

        delete_after_objects = []
        for h in reversed(l):
            ref = get_object_ref(h.object)
            if ref not in wait_results:
                continue
            do_delete = False
            if wait_results[ref] and "hook-succeeded" in h.delete_policies:
                do_delete = True
            elif not wait_results[ref] and "hook-failed" in h.delete_policies:
                do_delete = True
            if do_delete:
                delete_after_objects.append(h)

        if len(delete_after_objects) != 0:
            do_log(logging.INFO, "Deleting %d hooks after hook execution" % len(delete_after_objects))
        for h in delete_after_objects:
            do_log(logging.DEBUG, "Deleting hook %s due to hook-delete-policy %s" % (get_long_object_name(h.object), ",".join(h.delete_policies)))
            self.apply_util.delete_object(get_object_ref(h.object))

    def get_hook(self, o) -> Optional[Hook]:
        def get_list(path):
            s = get_dict_value(o, path, "")
            s = s.split(",")
            s = [x for x in s if x != ""]
            return s

        supported_kluctl_hooks = {"pre-deploy", "post-deploy",
                                  "pre-deploy-initial", "post-deploy-initial",
                                  "pre-deploy-upgrade", "post-deploy-upgrade"}
        supported_kluctl_delete_policies = {"before-hook-creation",
                                            "hook-succeeded",
                                            "hook-failed"}

        # delete/rollback hooks are actually not supported, but we won't show warnings about that to not spam the user
        supported_helm_hooks = {"pre-install", "post-install",
                                "pre-upgrade", "post-upgrade",
                                "pre-delete", "post-delete",
                                "pre-rollback", "post-rollback"}

        hooks = get_list('metadata.annotations."kluctl.io/hook"')
        for h in hooks:
            if h not in supported_kluctl_hooks:
                self.apply_util.handle_error(get_object_ref(o), "Unsupported kluctl.io/hook '%s'" % h)

        helm_hooks = get_list('metadata.annotations."helm.sh/hook"')
        for h in helm_hooks:
            if h not in supported_helm_hooks:
                self.apply_util.deployment_collection.add_api_warnings(get_object_ref(o), ["Unsupported helm.sh/hook '%s'" % h])

        if "pre-install" in helm_hooks:
            hooks.append("pre-deploy-initial")
        if "pre-upgrade" in helm_hooks:
            hooks.append("pre-deploy-upgrade")
        if "post-install" in helm_hooks:
            hooks.append("post-deploy-initial")
        if "post-upgrade" in helm_hooks:
            hooks.append("post-deploy-upgrade")
        if "pre-delete" in helm_hooks:
            hooks.append("pre-delete")
        if "post-delete" in helm_hooks:
            hooks.append("post-delete")
        hooks = set(hooks)

        weight = get_dict_value(o, 'metadata.annotations."kluctl.io/hook-weight"')
        if weight is None:
            weight = get_dict_value(o, 'metadata.annotations."helm.sh/hook-weight"')
        if weight is None:
            weight = "0"
        weight = int(weight)

        delete_policy = get_list('metadata.annotations."kluctl.io/hook-delete-policy"')
        delete_policy += get_list('metadata.annotations."helm.sh/hook-delete-policy"')
        if not delete_policy:
            delete_policy = ["before-hook-creation"]
        delete_policy = set(delete_policy)

        for p in delete_policy:
            if p not in supported_kluctl_delete_policies:
                self.apply_util.handle_error(get_object_ref(o), "Unsupported kluctl.io/hook-delete-policy '%s'" % p)

        try:
            wait = parse_bool(get_dict_value(o, 'metadata.annotations."kluctl.io/hook-wait"', "true"), do_raise=True)
        except ValueError as e:
            self.apply_util.handle_error(get_object_ref(o), str(e))
            wait = True

        if not hooks:
            return None

        return Hook(object=o, hooks=hooks, weight=weight, delete_policies=delete_policy, wait=wait)

    def get_sorted_hooks_list(self, l) -> List[Hook]:
        ret = []
        for x in l:
            h = self.get_hook(x)
            if h is None:
                continue
            ret.append(h)
        ret.sort(key=lambda x: x.weight)
        return ret

File: pkg/deployment/hooks_util.go
package deployment

import (
	"fmt"
	"github.com/codablock/kluctl/pkg/types"
	"github.com/codablock/kluctl/pkg/utils"
	log "github.com/sirupsen/logrus"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"sort"
	"strconv"
	"strings"
)

var supportedKluctlHooks = []string{
	"pre-deploy", "post-deploy",
	"pre-deploy-initial", "post-deploy-initial",
	"pre-deploy-upgrade", "post-deploy-upgrade",
}

var supportedKluctlDeletePolicies = []string{
	"before-hook-creation",
	"hook-succeeded",
	"hook-failed",
}

// delete/rollback hooks are actually not supported, but we won't show warnings about that to not spam the user
var supportedHelmHooks = []string{
	"pre-install", "post-install",
	"pre-upgrade", "post-upgrade",
	"pre-delete", "post-delete",
	"pre-rollback", "post-rollback",
}

type hooksUtil struct {
	a *applyUtil
}

type hook struct {
	object         *unstructured.Unstructured
	hooks          map[string]bool
	weight         int
	deletePolicies map[string]bool
	wait           bool
}

func (u *hooksUtil) runHooks(d *deploymentItem, hooks []string) {
	doLog := func(level log.Level, s string, f ...interface{}) {
		u.a.doLog(d, level, s, f...)
	}

	var l []*hook
	for _, h := range u.getSortedHooksList(d.objects) {
		for h2 := range h.hooks {
			if utils.FindStrInSlice(hooks, h2) != -1 {
				l = append(l, h)
				break
			}
		}
	}

	if len(l) != 0 && log.IsLevelEnabled(log.DebugLevel) {
		doLog(log.DebugLevel, "Sorted hooks:")
		for _, h := range l {
			doLog(log.DebugLevel, "  %s", types.RefFromObject(h.object).String())
		}
	}

	var deleteBeforeObjects []*hook
	var applyObjects []*hook

	for _, h := range l {
		if u.a.abortSignal {
			return
		}
		if _, ok := h.deletePolicies["before-hook-creation"]; ok {
			deleteBeforeObjects = append(deleteBeforeObjects, h)
		}
		applyObjects = append(applyObjects, h)
	}

	doDeleteForPolicy := func(h *hook) bool {
		ref := types.RefFromObject(h.object)
		var dpStr []string
		for p := range h.deletePolicies {
			dpStr = append(dpStr, p)
		}
		doLog(log.DebugLevel, "Deleting hook %s due to hook-delete-policy %s", ref.String(), strings.Join(dpStr, ","))
		return u.a.deleteObject(ref)
	}

	if len(deleteBeforeObjects) != 0 {
		doLog(log.InfoLevel, "Deleting %d hooks before hook execution", len(deleteBeforeObjects))
	}
	for _, h := range deleteBeforeObjects {
		if !doDeleteForPolicy(h) {
			return
		}
	}

	waitResults := make(map[types.ObjectRef]bool)
	if len(applyObjects) != 0 {
		doLog(log.InfoLevel, "Applying %d hooks", len(applyObjects))
	}
	for _, h := range applyObjects {
		ref := types.RefFromObject(h.object)
		_, replaced := h.deletePolicies["before-hook-creation"]
		doLog(log.DebugLevel, "Applying hook %s", ref.String())
		u.a.applyObject(h.object, replaced, true)

		if u.a.hadError(ref) {
			continue
		}
		if !h.wait {
			continue
		}
		waitResults[ref] = u.a.waitHook(ref)
	}

	var deleteAfterObjects []*hook
	for i := len(applyObjects) - 1; i >= 0; i-- {
		h := applyObjects[i]
		ref := types.RefFromObject(h.object)
		waitResult, ok := waitResults[ref]
		if !ok {
			continue
		}
		doDelete := false
		if _, ok := h.deletePolicies["hook-succeeded"]; waitResult && ok {
			doDelete = true
		} else if _, ok := h.deletePolicies["hook-failed"]; !waitResult && ok {
			doDelete = true
		}
		if doDelete {
			deleteAfterObjects = append(deleteAfterObjects, h)
		}
	}

	if len(deleteAfterObjects) != 0 {
		doLog(log.InfoLevel, "Deleting %d hooks after hook execution", len(deleteAfterObjects))
	}
	for _, h := range deleteAfterObjects {
		doDeleteForPolicy(h)
	}
}

func (u *hooksUtil) getHook(o *unstructured.Unstructured) *hook {
	ref := types.RefFromObject(o)
	getSet := func(name string) map[string]bool {
		ret := make(map[string]bool)
		a, ok := o.GetAnnotations()[name]
		if !ok {
			return ret
		}
		for _, x := range strings.Split(a, ",") {
			if x != "" {
				ret[x] = true
			}
		}
		return ret
	}

	hooks := getSet("kluctl.io/hook")
	for h := range hooks {
		if utils.FindStrInSlice(supportedKluctlHooks, h) == -1 {
			u.a.handleError(ref, fmt.Errorf("unsupported kluctl.io/hook '%s'", h))
		}
	}

	helmHooks := getSet("helm.sh/hook")
	for h := range helmHooks {
		if utils.FindStrInSlice(supportedHelmHooks, h) == -1 {
			u.a.handleError(ref, fmt.Errorf("unsupported helm.sh/hook '%s'", h))
		}
	}

	helmCompatibility := func(helmName string, kluctlName string) {
		if _, ok := helmHooks[helmName]; ok {
			hooks[kluctlName] = true
		}
	}

	helmCompatibility("pre-install", "pre-deploy-initial")
	helmCompatibility("pre-upgrade", "pre-deploy-upgrade")
	helmCompatibility("post-install", "post-deploy-initial")
	helmCompatibility("post-upgrade", "post-deploy-upgrade")
	helmCompatibility("pre-delete", "pre-delete")
	helmCompatibility("post-delete", "post-delete")

	weightStr, ok := o.GetAnnotations()["kluctl.io/hook-weight"]
	if !ok {
		weightStr, ok = o.GetAnnotations()["helm.sh/hook-weight"]
	}
	if !ok {
		weightStr = "0"
	}
	weight, err := strconv.ParseInt(weightStr, 10, 32)
	if err != nil {
		u.a.handleError(ref, fmt.Errorf("failed to parse hook weight: %w", err))
	}

	deletePolicy := getSet("kluctl.io/hook-delete-policy")
	for d := range getSet("helm.sh/hook-delete-policy") {
		deletePolicy[d] = true
	}
	if len(deletePolicy) == 0 {
		deletePolicy["before-hook-creation"] = true
	}

	for p := range deletePolicy {
		if utils.FindStrInSlice(supportedKluctlDeletePolicies, p) == -1 {
			u.a.handleError(ref, fmt.Errorf("unsupported kluctl.io/hook-delete-policy '%s'", p))
		}
	}

	waitStr, ok := o.GetAnnotations()["kluctl.io/hook-wait"]
	if !ok {
		waitStr = "true"
	}
	wait, err := strconv.ParseBool(waitStr)
	if err != nil {
		u.a.handleError(ref, fmt.Errorf("failed to parse %s as bool", waitStr))
		wait = true
	}

	if len(hooks) == 0 {
		return nil
	}

	return &hook{
		object:         o,
		hooks:          hooks,
		weight:         int(weight),
		deletePolicies: deletePolicy,
		wait:           wait,
	}
}

func (u *hooksUtil) getSortedHooksList(objects []*unstructured.Unstructured) []*hook {
	var ret []*hook
	for _, o := range objects {
		h := u.getHook(o)
		if h == nil {
			continue
		}
		ret = append(ret, h)
	}
	sort.SliceStable(ret, func(i, j int) bool {
		return ret[i].weight < ret[j].weight
	})
	return ret
}
